### ISSUE TITLE
DXE-2728 Import property and apply changes fails with: "Property name already in use"

### ISSUE THREAD

**Issue by mehanig** [akamai/terraform-provider-akamai#426]
Hello, 
I am having troubles importing property and making a change to it via terraform.
I might be missunderstanding some parts of the process, but can't figure out, what am I doing wrong.

Worth mentioning, that similar process worked for me about 6 months ago, but now it doesn't. I can see there were some updates in provider and rules versions, but not sure if I need to do anything about it.

Here is how I expected it to work from scratch:

1) have Akamai CLI + Terraform installed
2) Setup auth through `.edgerc` file
3) Create empty project with just 1 file, to import existing property (followed by this doc: https://techdocs.akamai.com/terraform/docs/set-up-property-provisioning#import-a-property)
4) Make changes to json file and run `terraform apply` to create new version of property

For Step 3, I import property configuration using `akamai` cli, and it works, synced configuration looks correct to me:
`akamai terraform --edgerc ~/.edgerc --section default export-property redacted_property`:
```
Fetching property redacted_property [=======   ] Fetching property redacted_property ... [OK]
Fetching group  [========  ] Fetching group  ... [OK]
Fetching property version  [=         ] Fetching property version  ... [OK]
Fetching property rules  [=====     ] Fetching property rules  ... [OK]
Fetching product name  [======    ] Fetching product name  ... [OK]
Fetching hostnames  [========  ] Fetching hostnames  ... [OK]
Fetching activation details  [========= ] Fetching activation details  ... [OK]
Saving TF configurations  ... [OK]
Terraform configuration for property 'redacted_property' was saved successfully
```
Terraform structure is created after that:
```
.
├── .terraform
│   └── providers
│       └── registry.terraform.io
│           └── akamai
│               └── akamai
│                   └── 4.1.0
│                       └── darwin_arm64
│                           ├── CHANGELOG.md
│                           ├── LICENSE
│                           ├── README.md
│                           └── terraform-provider-akamai_v4.1.0
├── .terraform.lock.hcl
├── import.sh
├── property-snippets
│   ├── Content_Compression.json
│   ├── Dynamic_Content.json
│   ├── Enable_Http2.json
│   ├── Failover_Origin.json
│   ├── New_Rule.json
│   ├── Static_Content.json
│   └── main.json
├── property.tf
├── terraform.tfstate
├── terraform.tfstate.backup
└── variables.tf
```

After this import, I expected `terraform plan -refresh-only` to show no changes, which it does:
```
data.akamai_group.group: Reading...
data.akamai_property_rules_template.rules: Reading...
data.akamai_property_rules_template.rules: Read complete after 0s [id=***]
data.akamai_group.group: Read complete after 1s [id=***]
data.akamai_contract.contract: Reading...
data.akamai_contract.contract: Read complete after 0s [id=***]

No changes. Your infrastructure still matches the configuration.

Terraform has checked that the real remote objects still match the result of your most recent changes, and found no differences.
```
**And here the problem starts:**
If I try to run `terraform apply` with the same configuration, even without making any changes, it fails with
**"title": "Property name already in use",**
```
akamai_property.redacted_property: Creating...
╷
│ Error: creating property: API error:
│ {
│ 	"type": "https://problems.luna.akamaiapis.net/papi/v0/property/name-in-use",
│ 	"title": "Property name already in use",
│ 	"detail": "The name `redacted_property` is already being used.",
│ 	"instance": "https://akab-my-instance.luna.akamaiapis.net/papi/v1/properties?contractId=***",
│ 	"statusCode": 400
│ }
│
│   with akamai_property.redacted_property,
│   on property.tf line 29, in resource "akamai_property" "redacted_property":
│   29: resource "akamai_property" "redacted_property" {
│
╵
```

Is it that I am missing some important step, so that akamai provider recognizes existing property configuration, and not tries to recreate it, or there is something else that I am doing wrong?

**Comment by dawiddzhafarov** [comment]
Hi @mehanig,

Thank You for opening the issue. First thing that comes into my mind, after executing `akamai terraform export-property` command, there should be 3 files and 1 directory created: `property.tf`, `variables.tf`, `import.sh` and property-snippets directory. After this step, the configuration was `exported`, not `imported`. In order to import the property resource, make sure to run `import.sh` script, i.e. via command `sh import.sh` (depending on environment You are using). This will actually import Your property to the state file (`terraform.tfstate`). Then You can proceed with any changes You want to apply to it.

The error You are encountering is caused by not importing the property to Your state file. It is empty, so after hitting `terraform apply` the resource tries to create new property, but fails as the property with given name already exists.

You have also listed files in Your working directory and there already is `terraform.tfstate` file - which could not be here, unless You have imported some resource or it was already there and contains different resources/data sources? 

Let me know if it solves Your issue.

With regards,
Dawid

**Comment by mehanig** [comment]
Hello @dawiddzhafarov,

Absolutelly appreciate taking a look at it.
I somehow completelly missed the part about executing `sh import.sh`, haven't seen it in docs, my bad.

Your suggestion partially fixed the issue, but still something isn't right about `contract_id`'s

So, I am performing those steps now:

0) create file `get_property.tf` with this content:
```
data "akamai_properties" "get-my-properties" {
  contract_id = "ctr_3-XXXXX"
  group_id    = "grp_XXXXX"
}

output "my-property-list" {
  value = data.akamai_properties.get-my-properties
}
```
1. **`akamai terraform --edgerc ~/.edgerc --section default export-property redacted_property`**
output:
```
Fetching property redacted_property [======    ] Fetching property redacted_property ... [OK]
Fetching group  [========  ] Fetching group  ... [OK]
Fetching property version  [=         ] Fetching property version  ... [OK]
Fetching property rules  [======    ] Fetching property rules  ... [OK]
Fetching product name  ... [OK]
Fetching hostnames  [========= ] Fetching hostnames  ... [OK]
Fetching activation details  [          ] Fetching activation details  ... [OK]
Saving TF configurations  ... [OK]
Terraform configuration for property 'redacted_property' was saved successfully
```

2) **`sh import.sh`**
```
Initializing the backend...

Initializing provider plugins...
- Finding akamai/akamai versions matching ">= 3.5.0"...
- Installing akamai/akamai v4.1.0...
- Installed akamai/akamai v4.1.0 (signed by a HashiCorp partner, key ID A26ECDD8F0BCBA73)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
data.akamai_group.group: Reading...
data.akamai_properties.get-my-properties: Reading...
data.akamai_property_rules_template.rules: Reading...
data.akamai_property_rules_template.rules: Read complete after 0s [id=***]
data.akamai_group.group: Read complete after 1s [id=grp_XXXX]
data.akamai_contract.contract: Reading...
data.akamai_contract.contract: Read complete after 0s [id=ctr_2-YYYYY]
akamai_property.redacted_property: Importing from ID "prp_***,ctr_3-XXXXX,grp_****,LATEST"...
akamai_property.redacted_property: Import prepared!
  Prepared akamai_property for import
akamai_property.redacted_property: Refreshing state... [id=prp_***]
data.akamai_properties.get-my-properties: Read complete after 2s [id=grp_***]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```


3) **`terraform plan`**
output:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
  ~ update in-place

Terraform will perform the following actions:

  # akamai_property.redacted_property will be updated in-place
  ~ resource "akamai_property" "redacted_property" {
      ~ contract_id        = "ctr_3-XXXXX" -> "ctr_2-YYYYY"
        id                 = "prp_*****"
        name               = "redacted_property"
        # (12 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }

  # akamai_property_activation.redacted_property will be created
  + resource "akamai_property_activation" "redacted_property" {
      + activation_id                  = (known after apply)
      + auto_acknowledge_rule_warnings = true
      + contact                        = [
          + "****",
        ]
      + errors                         = (known after apply)
      + id                             = (known after apply)
      + network                        = "STAGING"
      + note                           = "Demo purposes"
      + property                       = (known after apply)
      + property_id                    = "prp_****"
      + status                         = (known after apply)
      + version                        = 20
      + warnings                       = (known after apply)
    }

Plan: 1 to add, 1 to change, 0 to destroy.
```

So in short, for some reason, it tries to update contract_id of the resource, which will fail I assume.
In `import.sh` file, it uses correct Contract_id, the one that I can see in Akamai dashboard, but when I run `terraform plan` it for some reason tries to update it.
In tfstate file, that is being created after `import.sh`, I can see both contract_id's being used, but I am not recognizing `ctr_2-YYYYY`. original one - `ctr_3-XXXXX` I know.

Is there a way to check what is ctr_2-YYYYY somehow?
In tfstate I can see that contract_id for group is ctr_2-YYYYY, but contract_id for property is `ctr_3-XXXXX`.

I tried to create `get_property.tf` with ctr_2-YYYYY, but that didn't changed much.
I am probably again missing something, let me know if that's some obvious mistake that I made.

**Comment by dawiddzhafarov** [comment]
This issue is most probably caused by two existing groups with the same name, but different IDs (IDs are unique, names are not). The group with different contract `ctr_2` is found first and returned in the data source for groups, while the group assigned to Your contract for property has the same name. We are planning to add an error explaining this problem and suggesting changing the name of Your group to be unique.

To fix this issue, I suggest trying to rename Your group so it will be unique. Once it's done, please perform the same steps You have described and let me know if it solves Your issue.

With regards,
Dawid

**Comment by mehanig** [comment]
Hello @dawiddzhafarov 

I think you might be right about same properties in different contract, but I can't see it. Maybe it's hidden or archived.

Let me explain what I tried:

1) Cloned problematic property with new name to `ctr_3` (our active contract) -- That didn't solved the issue for new property, I was still getting terraform configuration trying to switch to`ctr_2`
2) Cloned problematic property with new name to `ctr_2` (our old contract I believe) -- That worked and I can make changes to property using Terraform provider and everything looks good.

I am experimenting my changes in "playground" profile, which has no traffic, so I am currenty not blocked and can prototype managing Akamai via Terraform, but this will not work for production workload, where we still need to use `ctr_2`

**Comment by dawiddzhafarov** [comment]
Hi @mehanig,

If I understood correctly, You have changed the property name to be different, right? I believe that the problem lies in the property's `group` name. The `akamai_contract` data source which is also exported while property is exported, uses `group_name` to find the corresponding group. Groups can have duplicate names, so because of that, the found group might not be the one expected. Make sure to try rename the `group` instead of the property.

At last, as a temporary workaround You could try replacing:
```
data "akamai_contract" "contract" {
  group_name = data.akamai_group.group.group_name
}
```
with:
```
data "akamai_contract" "contract" {
  group_id = data.akamai_group.group.id
}
```
Let me know if that helps.

With regards,
Dawid

**Comment by lkowalsk-akamai-com** [comment]
This issue should be resolved with recent release of cli-terraform.

**Comment by mehanig** [comment]
Perfect, I verified that it now works without any extra steps.
Thanks for a fix!
